pkg/common/cns-lib/vsphere/virtualcenter.go

```diff
@@ -228,11 +231,23 @@ func (vc *VirtualCenter) connect(ctx context.Context, requestNewSession bool) er
 		sessionMgr := session.NewManager(vc.Client.Client)
 		// SessionMgr.UserSession(ctx) retrieves and returns the SessionManager's CurrentSession field
 		// Nil is returned if the session is not authenticated or timed out.
-		if userSession, err := sessionMgr.UserSession(ctx); err != nil {
+		var userSession *types.UserSession
+		if userSession, err = sessionMgr.UserSession(ctx); err != nil {
 			log.Errorf("failed to obtain user session with err: %v", err)
 			return err
-		} else if userSession != nil {
-			return nil
+		}
+		if userSession != nil {
+			activeSession, err := sessionMgr.SessionIsActive(ctx)
+			if err != nil {
+				log.Errorf("failed to check user session is active. err: %v", err)
+				return err
+			}
+			if !activeSession {
+				return errors.New("session is not active")
```

Comment:
How does the system recover in this case?

Reply:
system recovers by calling vc.connect(ctx, true). Note here second argument requestNewSession is set to true.

```
// Connect establishes a new connection with vSphere with updated credentials
// If credentials are invalid then it fails the connection.
func (vc *VirtualCenter) Connect(ctx context.Context) error {
	log := logger.GetLogger(ctx)
	// Set up the vc connection
	err := vc.connect(ctx, false)
	if err != nil {
		log.Errorf("Cannot connect to vCenter with err: %v", err)
		log.Infof("attempting to reconnect with requesting a new session")
		return vc.connect(ctx, true)
	}
	return nil
}
```

Reply:
How are we making sure that concurrent requests to Connect method is creating a single session? With the changed code I see that many threads may call vc.connect(ctx, true) and end up creating new sessions and clients.

Reply:
we are protected by clientMutex in private connect method.

```
// connect creates a connection to the virtual center host.
func (vc *VirtualCenter) connect(ctx context.Context, requestNewSession bool) error {
	log := logger.GetLogger(ctx)
	clientMutex.Lock()
	defer clientMutex.Unlock()
```

Reply:
I got what you are saying. let me see, how we can improve on this.

Reply:
@SandeepPissay addressed the concern of driver end up creating multiple session with this commit - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/503/commits/f33b3355617fb593684804cb339878c39e42ae9b, by removing the usage of requestNewSession argument, and made sure all thread uses single lock `clientMutex` and local variable `requestNewSession` which we are initially setting to false and then determining to create a new session depending on the result of the checks.